### ISSUE TITLE
Potentially fatal error in state experiments

### ISSUE THREAD

**Issue by ellen-ejkim** [JonathanCrabbe/Dynamask#4]
Dear authors,

In the experiments with the state dataset,
the output shape of the model appears to be: [(B x n_state), T] (B: batch_size, n_state: number of classes, T: the number of time steps)
(Because "return_all=True").

Then, in line 52 of "experiments/state.py" you use the reshape function to change the output shape into [T, n_state].
> def f(x):
>        x = x.unsqueeze(0)
>        x = x.transpose(1, 2)
>        out = model(x)
>        out = out.reshape(T, -1)
>        out = softmax(out)
>        return out

In the current form (with the reshape function), the softmax function is not actually calculating the class probability, since it is applied between two time-steps instead of classes. 
Therefore, the output of this function f(x) is not a correct prediction output of the model.

We believe that to correctly reshape the output tensor into a desired shape, the transpose function should be used instead of the reshape function. ---> ex) a= [[1, 2, 3], [4, 5, 6]] a.tranpose(): [[1, 4], [2, 5], [3, 6]] a.reshape() = [[1, 2], [3, 4], [5, 6]]
> out = out.transpose(0, 1)

When the reshape function is replaced with the transpose function, the explanation results get completely destroyed.

If our inspection of your implementation is correct, this could significantly undermine the credibility of the reported results.
We strongly recommend that you double-check your implementation; please let us know of what you think.

**Comment by JonathanCrabbe** [comment]
Hello,

Dynamask is a perturbation based approach that studies the sensitivity of a black-box to perturbations. If you pay attention to the implementation, you'll notice that the mask optimization only involves f(x_pert) and f(x), where x_pert is a perturbed version of the input x. It is crucial to understand that no ground truth label are involved in this optimization: only a black-box sensitivity analysis. This leaves you with a lot of freedom on what you call _the black-box output_.

Now let us focus on the implementation of the state experiment. As you rightfully said, the optimization involves a comparison between the predicted probabilities for a given class at two consecutive time steps. Increasing the log loss between these two probabilities has the effect of increasing one while decreasing the other. In this way, the error part of the objective function helps to find the perturbation that maximally contrasts the predicted probability for a given class from one time step to the next one. I believe that this approach is valid and this is confirmed by the very good results. 

If you insist on matching the perturbed and unperturbed class probabilities at a given time step, it makes much more sense to use the cross-entropy rather than the log-loss. If you do this, you'll reach essentially the same conclusions as in the paper (Dynamask has the best precision and the second best recall) although the results will not be as good as with the above approach. If you want to try this, you can simply modify the relevant parts of the code accordingly:

```python

from utils.losses import cross_entropy

(...)

  # The black-box is defined with the model:
    def f(x):
        x = x.unsqueeze(0)
        x = x.transpose(1, 2)
        out = model(x)
        out = out.transpose(1, 0)
        out = softmax(out)
        return out

         (...)        

        # Fit the group of mask:
        mask_group = MaskGroup(pert, device, verbose=False, deletion_mode=False)
        mask_group.fit(X=x_test, f=f, loss_function=cross_entropy, area_list=area_list,
                       learning_rate=1.0, size_reg_factor_init=0.1, size_reg_factor_dilation=100,
                       initial_mask_coeff=0.5, n_epoch=1500, momentum=0.9, time_reg_factor=1.0)  
        mask = mask_group.get_best_mask()

         (...)        
````

A smarter approach is to match the model's output in pre-activation space. This yields results that are closer to the one obtained with the first approach. If you want to try this, you can simply modify the relevant parts of the code accordingly:

```python

from utils.losses import mse

(...)

  # The black-box is defined with the model:
    def f(x):
        x = x.unsqueeze(0)
        x = x.transpose(1, 2)
        out = model(x)
        return out

         (...)        

        # Fit the group of mask:
        mask_group = MaskGroup(pert, device, verbose=False, deletion_mode=False)
        mask_group.fit(X=x_test, f=f, loss_function=mse, area_list=area_list,
                       learning_rate=1.0, size_reg_factor_init=0.1, size_reg_factor_dilation=100,
                       initial_mask_coeff=0.5, n_epoch=1500, momentum=0.9, time_reg_factor=1.0)  
        mask = mask_group.get_best_mask()

         (...)        
````

I believe that you could obtain even better results by matching in the RNN's latent space but I leave this to you as an exercice. 

In any case, you can rest assured: there is no _fatal error_ nor results that get completely _destroyed_. A friendly advice on this note: it is wise to avoid such terminology when you interact with researchers. This does not bring anything to the message you want to convey and makes it sound quite unfriendly. 

Best,
Jonathan

**Comment by ellen-ejkim** [comment]
Dear authors,

I apologize if I came off as a bit rude in my last post.

However, it appears that there is some misalignment between the content of the paper and the actual implementation. Please correct me if I’m wrong, but in the main paper, I believe you do not explicitly state that Dynamask compares the predicted probabilities of two consecutive time steps. Because the choice of probabilities and how they are used to calculate the loss are central to understanding the intent of the proposed method, leaving out this detail could confuse future researchers.
Once again, please let me know if this explanation is already included in the paper.

Before my last post, I used “log_loss,” which yielded quite disappointing results (hence the word destroyed). I confirmed that using  “cross_entropy” obtains better results. 

I still don't understand the logic behind comparing two time steps, but I'll think about this more. I do not think “very good results” can solely support the validity of such an approach. Within the paper in its current form, it is difficult to find motivation or theoretical support for this part.
